File: day1/day1.cpp
#include <fstream>
#include <unordered_map>

#include <fmt/format.h>

using namespace std;

struct Pos
{
	int x;
	int y;

	size_t manhattan() const noexcept
	{
		return (x>0?x:-x) + (y>0?y:-y);
	}
};

static bool operator==(Pos a, Pos b)
{
	return a.x == b.x && a.y == b.y;
}

static Pos operator+(Pos a, Pos b)
{
	return Pos{a.x+b.x, a.y+b.y};
}

static Pos xmul(Pos a, Pos b)
{
	return Pos{a.x*b.x - a.y*b.y, a.x*b.y + a.y*b.x};
}

struct HashPos
{
	size_t operator()(const Pos& p) const noexcept
	{
		return (hash<int>{}(p.x)<<4)^(hash<int>{}(p.y));
	}
};

int main(int argc, char *argv[])
{
	if (argc < 2)
	{
		fmt::print(stderr, "Usage: {} <filename>\n", argv[0]);
		return 1;
	}

	ifstream input(argv[1]);
	if (!input)
	{
		fmt::print(stderr, "Cannot open {}\n", argv[1]);
		return 1;
	}

	string directions;
	getline(input, directions);
	input.close();
	if(input.fail())
	{
		fmt::print(stderr, "Cannot parse the data\n");
		return 1;
	}

	unordered_map<Pos, bool, HashPos> visited;
	int twice = 0;
	Pos cur{0, 0};
	Pos dir{0, -1};
	visited[cur] = true;
	size_t pos = 0;
	size_t next;
	do
	{
		next = directions.find(' ', pos);
		int v = directions[pos];
		int l = stoi(directions.substr(pos+1, next-1));
		dir = xmul(dir, v == 'R' ? Pos{0,1} : Pos{0,-1});
		while (l-->0)
		{
			cur = cur + dir;
			if (twice == 0 && visited.find(cur) != visited.end())
			{
				twice = cur.manhattan();
			}
			visited[cur] = true;
		}
		pos = next + 1;
	} while (next != string::npos);

	fmt::print("Part1: {}\nPart2: {}\n", cur.manhattan(), twice);
	return 0;
}

File: day1/day1.py
#!/usr/bin/env python3

def add(a, b):
    return (a[0]+b[0], a[1]+b[1])

def xmul(a, b):
    return (a[0]*b[0]-a[1]*b[1], a[0]*b[1]+a[1]*b[0])

def manhattan(a):
    return abs(a[0]) + abs(a[1])

def walk(cmds):
    pos = (0, 0)
    dir = (0, -1)
    for cmd in cmds.split(", "):
        if cmd[0] == "R":
            dir = xmul(dir, (0, 1))
        else:
            dir = xmul(dir, (0, -1))

        for i in range(int(cmd[1:])):
            pos = add(pos, dir)

    return manhattan(pos)

def twice(cmds):
    map = {}
    pos = (0, 0)
    dir = (0, -1)
    for cmd in cmds.split(", "):
        if cmd[0] == "R":
            dir = xmul(dir, (0, 1))
        else:
            dir = xmul(dir, (0, -1))

        for i in range(int(cmd[1:])):
            pos = add(pos, dir)
            if pos in map:
                return manhattan(pos)
            map[pos] = 1


assert walk("R2, L3") == 5
assert walk("R2, R2, R2") == 2
assert walk("R5, L5, R5, R3") == 12
assert twice("R8, R4, R4, R8") == 4

import sys
if __name__ == "__main__":
    if len(sys.argv) < 2:
        print("Usage: {} <filename>".format(sys.argv[0]), file=sys.stderr)
        sys.exit(1)

    try:
        with open(sys.argv[1], "rt") as f:
            txt = f.read().strip()
    except:
        print("Cannot open {}".format(sys.argv[1]), file=sys.stderr)
        sys.exit(1)

    print("Part1:", walk(txt))
    print("Part2:", twice(txt))
